File: youtube/lat5-wine/try1.py
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns 
from statsmodels.stats.outliers_influence import variance_inflation_factor
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_selection import RFE
import statsmodels.api as sm 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor

df = pd.read_csv('ML/youtube/lat5-wine/WineQT.csv')
print(df.head())
print(df.info())
print(df.describe())
print(df.isnull().sum())

# fig, ax = plt.subplots(3,4, figsize=(10,5))
# plt1 = sns.boxplot(df['fixed acidity'], ax = ax[0,0])
# plt2 = sns.boxplot(df['volatile acidity'], ax = ax[0,1])
# plt3 = sns.boxplot(df['citric acid'], ax = ax[0,2])
# plt4 = sns.boxplot(df['residual sugar'], ax = ax[0,3])
# plt5 = sns.boxplot(df['chlorides'], ax = ax[1,0])
# plt6 = sns.boxplot(df['free sulfur dioxide'], ax = ax[1,1])
# plt7 = sns.boxplot(df['total sulfur dioxide'], ax = ax[1,2])
# plt8 = sns.boxplot(df['density'], ax = ax[1,3])
# plt9 = sns.boxplot(df['pH'], ax = ax[2,0])
# plt10 = sns.boxplot(df['sulphates'], ax = ax[2,1])
# plt11 = sns.boxplot(df['alcohol'], ax = ax[2,2])
# plt12 = sns.boxplot(df['quality'], ax = ax[2,3])
# plt.tight_layout()
# plt.show()

# fig, ax = plt.subplots(3, 4, figsize=(15, 10))
# sns.scatterplot(x=df.index, y=df['fixed acidity'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['volatile acidity'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=df.index, y=df['citric acid'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['residual sugar'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=df.index, y=df['chlorides'], ax=ax[1, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['free sulfur dioxide'], ax=ax[1, 1], color='green')
# sns.scatterplot(x=df.index, y=df['total sulfur dioxide'], ax=ax[1, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['density'], ax=ax[1, 3], color='red')
# sns.scatterplot(x=df.index, y=df['pH'], ax=ax[2, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['sulphates'], ax=ax[2, 1], color='green')
# sns.scatterplot(x=df.index, y=df['alcohol'], ax=ax[2, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['quality'], ax=ax[2, 3], color='red')
# plt.tight_layout()
# plt.show()

q1 = df['fixed acidity'].quantile(0.25)
q3 = df['fixed acidity'].quantile(0.75)
iqr = q3 - q1
df = df[(df['fixed acidity'] >= q1 - 1.5*iqr) & (df['fixed acidity'] <= q3 + 1.5*iqr )]


q1 = np.percentile(df['volatile acidity'], 25)
q3 = np.percentile(df['volatile acidity'], 75)
iqr = q3 - q1
df = df[(df['volatile acidity'] >= q1 - 1.5*iqr) & (df['volatile acidity'] <= q3 + 1.5*iqr )]

q1 = df['citric acid'].quantile(0.25)
q3 = df['citric acid'].quantile(0.75)
iqr = q3 - q1
df = df[(df['citric acid'] >= q1 - 1.5*iqr) & (df['citric acid'] <= q3 + 1.5*iqr )]


q1 = np.percentile(df['residual sugar'], 25)
q3 = np.percentile(df['residual sugar'], 75)
iqr = q3 - q1
df = df[(df['residual sugar'] >= q1 - 1.5*iqr) & (df['residual sugar'] <= q3 + 1.5*iqr )]

q1 = df['chlorides'].quantile(0.25)
q3 = df['chlorides'].quantile(0.75)
iqr = q3 - q1
df = df[(df['chlorides'] >= q1 - 1.5*iqr) & (df['chlorides'] <= q3 + 1.5*iqr )]


q1 = np.percentile(df['free sulfur dioxide'], 25)
q3 = np.percentile(df['free sulfur dioxide'], 75)
iqr = q3 - q1
df = df[(df['free sulfur dioxide'] >= q1 - 1.5*iqr) & (df['free sulfur dioxide'] <= q3 + 1.5*iqr )]


q1 = np.percentile(df['total sulfur dioxide'], 25)
q3 = np.percentile(df['total sulfur dioxide'], 75)
iqr = q3 - q1
df = df[(df['total sulfur dioxide'] >= q1 - 1.5*iqr) & (df['total sulfur dioxide'] <= q3 + 1.5*iqr )]


q1 = df['density'].quantile(0.25)
q3 = df['density'].quantile(0.75)
iqr = q3 - q1
df = df[(df['density'] >= q1 - 1.5*iqr) & (df['density'] <= q3 + 1.5*iqr )]


q1 = df['pH'].quantile(0.25)
q3 = df['pH'].quantile(0.75)
iqr = q3 - q1
df = df[(df['pH'] >= q1 - 1.5*iqr) & (df['pH'] <= q3 + 1.5*iqr )]


q1 = df['sulphates'].quantile(0.25)
q3 = df['sulphates'].quantile(0.75)
iqr = q3 - q1
df = df[(df['sulphates'] >= q1 - 1.5*iqr) & (df['sulphates'] <= q3 + 1.5*iqr )]


q1 = df['alcohol'].quantile(0.25)
q3 = df['alcohol'].quantile(0.75)
iqr = q3 - q1
df = df[(df['alcohol'] >= q1 - 1.5*iqr) & (df['alcohol'] <= q3 + 1.5*iqr )]


# fig, ax = plt.subplots(3, 4, figsize=(15, 10))
# sns.scatterplot(x=df.index, y=df['fixed acidity'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['volatile acidity'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=df.index, y=df['citric acid'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['residual sugar'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=df.index, y=df['chlorides'], ax=ax[1, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['free sulfur dioxide'], ax=ax[1, 1], color='green')
# sns.scatterplot(x=df.index, y=df['total sulfur dioxide'], ax=ax[1, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['density'], ax=ax[1, 3], color='red')
# sns.scatterplot(x=df.index, y=df['pH'], ax=ax[2, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['sulphates'], ax=ax[2, 1], color='green')
# sns.scatterplot(x=df.index, y=df['alcohol'], ax=ax[2, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['quality'], ax=ax[2, 3], color='red')
# plt.tight_layout()
# plt.show()

del df['Id']

print(df.head())

scaler = MinMaxScaler()
numvars = df.drop('quality', axis = 1).columns
print(numvars)
df[numvars] = scaler.fit_transform(df[numvars])
print(df.head())

df_train, df_test = train_test_split(df, train_size = 0.7, test_size=0.3, random_state=100)

corr = df_train.corr()
print(corr)
# plt.figure(figsize=(16, 10))
# sns.heatmap(corr, annot = True, cmap="YlGnBu")
# plt.show()


x_train = df_train
y_train = df_train.pop('quality')

k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelNB  = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(modelNB, x_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelNB.fit(x_train, y_train)


rfe = RFE(estimator=modelNB, n_features_to_select=5)
rfe = rfe.fit(x_train, y_train)
print(rfe.support_)
print(rfe.ranking_)

hasil = pd.DataFrame({
    'features': x_train.columns,
    'support': rfe.support_,
    'rank': rfe.ranking_,
})
print(hasil)

support = x_train.columns[rfe.support_]
not_support = x_train.columns[~rfe.support_]
print(support)
print(not_support)

x_train_rfe = x_train[support]


k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelNB  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelNB, x_train_rfe, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelNB.fit(x_train_rfe, y_train)

vif = pd.DataFrame()
x = x_train_rfe
print(x.shape)

vif['features'] = x.columns
vif['VIF'] = [variance_inflation_factor(x.values, i) for i in range(x.shape[1])]
vif = vif.sort_values('VIF', ascending=False)
print(vif)

vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values('VIF', ascending=False)
print(vif)


y_pred = modelNB.predict(x_train_rfe)

df = pd.DataFrame({
    'y_train': y_train,
    'y_pred': y_pred,
})
print(df.head())

akurasi = accuracy_score(y_train, y_pred)
print(akurasi)
cm = confusion_matrix(y_train, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

x_test = df_test
y_test = df_test.pop('quality')
x_test_rfe = x_test[support]
y_pred = modelNB.predict(x_test_rfe)

df = pd.DataFrame({
    'y_test': y_test,
    'y_pred': y_pred,
})
print(df.head())

akurasi = accuracy_score(y_test, y_pred)
print(akurasi)
cm = confusion_matrix(y_test, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

File: youtube/lat5-wine/try1.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(dplyr)
library(tidyr)
library(plyr)
library(lubridate)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees

df <- read.csv("ML/youtube/lat5-wine/WineQT.csv")
print(head(df))
print(dim(df))
print(colSums(is.na(df)))
cat("\n\n")

df <- dplyr::select(df, -Id)
print(head(df))


# Buat plot boxplot untuk setiap variabel
boxplot1 <- ggplot(df, aes(x = "", y = `fixed.acidity`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot2 <- ggplot(df, aes(x = "", y = `volatile.acidity`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot3 <- ggplot(df, aes(x = "", y = `citric.acid`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot4 <- ggplot(df, aes(x = "", y = `residual.sugar`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot5 <- ggplot(df, aes(x = "", y = `chlorides`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot6 <- ggplot(df, aes(x = "", y = `free.sulfur.dioxide`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot7 <- ggplot(df, aes(x = "", y = `total.sulfur.dioxide`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot8 <- ggplot(df, aes(x = "", y = `density`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot9 <- ggplot(df, aes(x = "", y = `pH`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot10 <- ggplot(df, aes(x = "", y = `sulphates`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot11 <- ggplot(df, aes(x = "", y = `alcohol`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
boxplot12 <- ggplot(df, aes(x = "", y = `quality`)) +
    geom_boxplot() +
    theme_minimal() +
    labs(x = NULL)
# Menggabungkan plot menggunakan grid.arrange
# plot <- grid.arrange(boxplot1, boxplot2, boxplot3, boxplot4, ncol = 2)
plot <- boxplot1 + boxplot2 + boxplot3 + boxplot4 + boxplot5 + boxplot6 + boxplot7 + boxplot8 + boxplot9 + boxplot10 + boxplot11 + boxplot12
print(plot)



p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fixed.acidity), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = volatile.acidity), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = citric.acid), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = residual.sugar), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = chlorides), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = free.sulfur.dioxide), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = total.sulfur.dioxide), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = density), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = pH), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = sulphates), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = alcohol), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = quality), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
plot <- p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + p10 + p12
# plot <- (p1 | p2) / (p3 | p4)
print(plot)


q1 <- quantile(df$fixed.acidity, 0.25)
q3 <- quantile(df$fixed.acidity, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, fixed.acidity >= q1 - 1.5 * iqr & fixed.acidity <= q3 + 1.5 * iqr)

q1 <- quantile(df$volatile.acidity, 0.25)
q3 <- quantile(df$volatile.acidity, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
# df <- filter(df, volatile.acidity >= q1 - 1.5 * iqr & volatile.acidity <= q3 + 1.5 * iqr)
df <- filter(df, volatile.acidity >= q1 - 1.5 * iqr , volatile.acidity <= q3 + 1.5 * iqr)


q1 <- quantile(df$citric.acid, 0.25)
q3 <- quantile(df$citric.acid, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- df[(df$citric.acid >= q1 - 1.5 * iqr & df$citric.acid <= q3 + 1.5 * iqr),]

q1 <- quantile(df$residual.sugar, 0.25)
q3 <- quantile(df$residual.sugar, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, residual.sugar >= q1 - 1.5 * iqr & residual.sugar <= q3 + 1.5 * iqr)

q1 <- quantile(df$chlorides, 0.25)
q3 <- quantile(df$chlorides, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
# df <- filter(df, chlorides >= q1 - 1.5 * iqr & chlorides <= q3 + 1.5 * iqr)
df <- filter(df, chlorides >= q1 - 1.5 * iqr , chlorides <= q3 + 1.5 * iqr)


q1 <- quantile(df$free.sulfur.dioxide, 0.25)
q3 <- quantile(df$free.sulfur.dioxide, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
# df <- filter(df, free.sulfur.dioxide >= q1 - 1.5 * iqr & free.sulfur.dioxide <= q3 + 1.5 * iqr)
df <- filter(df, free.sulfur.dioxide >= q1 - 1.5 * iqr , free.sulfur.dioxide <= q3 + 1.5 * iqr)


q1 <- quantile(df$total.sulfur.dioxide, 0.25)
q3 <- quantile(df$total.sulfur.dioxide, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, total.sulfur.dioxide >= q1 - 1.5 * iqr & total.sulfur.dioxide <= q3 + 1.5 * iqr)


q1 <- quantile(df$density, 0.25)
q3 <- quantile(df$density, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, density >= q1 - 1.5 * iqr & density <= q3 + 1.5 * iqr)


q1 <- quantile(df$pH, 0.25)
q3 <- quantile(df$pH, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, pH >= q1 - 1.5 * iqr & pH <= q3 + 1.5 * iqr)


q1 <- quantile(df$sulphates, 0.25)
q3 <- quantile(df$sulphates, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, sulphates >= q1 - 1.5 * iqr & sulphates <= q3 + 1.5 * iqr)


q1 <- quantile(df$alcohol, 0.25)
q3 <- quantile(df$alcohol, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, alcohol >= q1 - 1.5 * iqr & alcohol <= q3 + 1.5 * iqr)


q1 <- quantile(df$quality, 0.25)
q3 <- quantile(df$quality, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, quality >= q1 - 1.5 * iqr & quality <= q3 + 1.5 * iqr)



p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fixed.acidity), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = volatile.acidity), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = citric.acid), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = residual.sugar), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = chlorides), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = free.sulfur.dioxide), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = total.sulfur.dioxide), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = density), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = pH), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = sulphates), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = alcohol), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = quality), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
# plot <- p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + p10 + p12
plot <- (p1 | p2 |p3) / (p4 | p5 | p6) / (p7|p8|p9) / (p10|p11|p12)
print(plot)

print(head(df))
print(dim(df))

cat('\n\n\n\n\n')

# Membuat objek scaler
scaler <- preProcess(dplyr::select(df, -quality), method = "range")
numvars <- colnames(dplyr::select(df, -quality))
df[numvars] <- predict(scaler, df[numvars])
print(head(df))

x <- dplyr::select(df, -quality)
y <- factor(df$quality)

set.seed(42)
index <- createDataPartition(y, p = 0.7, list = TRUE)
print(head(index))


x_train <- x[index$Resample1, ]
x_test <- x[-index$Resample1, ]
y_train <- y[index$Resample1]
y_test <- y[-index$Resample1]


ctrl <- trainControl(method = "cv", number = 5)
modelrf <- train(x_train, y_train, method = "rf", trControl = ctrl)
print("accuracy mean")
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat("\n")

# model <- train(x_train, y_train, method="nnet", trControl=ctrl)
# print('accuracy mean')
# print(model$result)
# print(mean(model$result$Accuracy))
# cat('\n')

# model <- train(x_train, y_train, method="rpart", trControl=ctrl)
# print('accuracy mean')
# print(model$result)
# print(mean(model$result$Accuracy))
# cat('\n')

# model <- train(x_train, y_train, method="nb", trControl=ctrl)
# print('accuracy mean')
# print(model$result)
# print(mean(model$result$Accuracy))
# cat('\n')

y_train_pred <- predict(modelrf, x_train)
print(y_train_pred)

df_pred <- data.frame(
    y_train = y_train,
    y_train_pred = y_train_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_train != y_train_pred)
print(head(df_pred))
# Calculate accuracy
accuracy <- confusionMatrix(data = y_train_pred, reference = y_train)$overall["Accuracy"]
print(accuracy)

cat("\n\n\n\n")


y_test_pred <- predict(modelrf, x_test)
print(y_test_pred)

df_pred <- data.frame(
    y_test = y_test,
    y_test_pred = y_test_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_test != y_test_pred)
print(head(df_pred))
# Calculate accuracy
accuracy <- confusionMatrix(data = y_test_pred, reference = y_test)$overall["Accuracy"]
print(accuracy)

cat("\n\n\n\n")
# Calculate confusion matrix
print("Confusion Matrix")
cm <- confusionMatrix(data = y_test_pred, reference = y_test)
print(cm)
